drivers/bluetooth/hci/hci_nxp.c

```diff
@@ -0,0 +1,1178 @@
+
+#define HCI_UART_CHILD_PREFIX(x) x
+#define HCI_UART_CHILD           DT_FOREACH_CHILD(DT_CHOSEN(zephyr_bt_uart), HCI_UART_CHILD_PREFIX)
+#if DT_NODE_HAS_PROP(HCI_UART_CHILD, sdio_reset_gpios)
+struct gpio_dt_spec sido_reset = GPIO_DT_SPEC_GET(HCI_UART_CHILD, sdio_reset_gpios);
+#endif /* DT_NODE_HAS_PROP(HCI_UART_CHILD, sdio_reset_gpios) */
+#if DT_NODE_HAS_PROP(HCI_UART_CHILD, w_disable_gpios)
+struct gpio_dt_spec w_disable = GPIO_DT_SPEC_GET(HCI_UART_CHILD, w_disable_gpios);
+#endif /* DT_NODE_HAS_PROP(HCI_UART_CHILD, w_disable_gpios) */
+
+struct nxp_ctlr_dev_data {
+	uint32_t primary_speed;
+	bool primary_flowcontrol;
+	uint32_t secondary_speed;
+	bool secondary_flowcontrol;
+};
+
+static struct nxp_ctlr_dev_data uart_dev_data;
+
+#define DI         0x07U
+#define POLYNOMIAL 0x04c11db7UL
+
+#define CRC32_LEN 4
+
+static unsigned char crc8_table[256U]; /* 8-bit table */
+static unsigned long crc_table[256U];
+static bool made_table;
+
+static void fw_upload_initCrc8(void)
+{
+	int i, j;
+	int crc;
+
+	if (!made_table) {
+		for (i = 0; i < 256; i++) {
+			crc = i;
+			for (j = 0; j < 8; j++) {
+				crc = (crc << 1) ^ ((crc & 0x80) ? DI : 0);
+			}
+			crc8_table[i] = (unsigned char)((unsigned char)crc & (unsigned char)0xFF);
+		}
+		made_table = true;
+	}
+}
+
+static void fw_upload_genCrcTable(void)
+{
+	int i, j;
+	unsigned long crc_accum;
+
+	for (i = 0; i < 256; i++) {
+		crc_accum = ((unsigned long)i << 24);
+		for (j = 0; j < 8; j++) {
+			if (crc_accum & 0x80000000L) {
+				crc_accum = (crc_accum << 1) ^ POLYNOMIAL;
+			} else {
+				crc_accum = (crc_accum << 1);
+			}
+		}
+		crc_table[i] = crc_accum;
+	}
+}
+
+static unsigned char fw_upload_crc8(unsigned char *array, unsigned char len)
+{
+	unsigned char CRC = 0xff;
+
+	for (; len > 0; len--) {
+		CRC = crc8_table[CRC ^ *array];
+		array++;
+	}
+	return CRC;
+}
+
+static unsigned long fw_upload_updateCrc(unsigned long crc_accum, char *data_blk_ptr,
+					 int data_blk_size)
+{
+	unsigned int i, j;
+
+	for (j = 0; j < data_blk_size; j++) {
+		i = ((unsigned int)(crc_accum >> 24) ^ *data_blk_ptr++) & 0xff;
+		crc_accum = (crc_accum << 8) ^ crc_table[i];
+	}
+	return crc_accum;
+}
+
+#define CMD4 0x4U
+#define CMD6 0x6U
+#define CMD7 0x7U
+
+#define V1_HEADER_DATA_REQ  0xa5U
+#define V1_START_INDICATION 0xaaU
+#define V1_REQUEST_ACK      0x5aU
+
+#define V3_START_INDICATION 0xabU
+#define V3_HEADER_DATA_REQ  0xa7U
+#define V3_REQUEST_ACK      0x7aU
+#define V3_TIMEOUT_ACK      0x7bU
+#define V3_CRC_ERROR        0x7cU
+
+#define REQ_HEADER_LEN    1U
+#define A6REQ_PAYLOAD_LEN 8U
+#define AbREQ_PAYLOAD_LEN 3U
+
+#define CRC_ERR_BIT          1 << 0U
+#define NAK_REC_BIT          1 << 1U
+#define TIMEOUT_REC_ACK_BIT  1 << 2U
+#define TIMEOUT_REC_HEAD_BIT 1 << 3U
+#define TIMEOUT_REC_DATA_BIT 1 << 4U
+#define INVALID_CMD_REC_BIT  1 << 5U
+#define WIFI_MIC_FAIL_BIT    1 << 6U
+#define BT_MIC_FAIL_BIT      1 << 7U
```

Comment:
these could probably use the `BIT()` macro instead

Reply:
Updated.